Number_types/include/CGAL/FPU.h

```diff
@@ -347,16 +346,32 @@ inline double IA_bug_sqrt(double d)
 // With GCC, we can do slightly better : test with __builtin_constant_p()
 // that both arguments are constant before stopping one of them.
 // Use inline functions instead ?
-#define CGAL_IA_ADD(a,b) CGAL_IA_FORCE_TO_DOUBLE((a)+CGAL_IA_STOP_CPROP(b))
-#define CGAL_IA_SUB(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)-(b))
-#define CGAL_IA_MUL(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)*CGAL_IA_STOP_CPROP(b))
-#define CGAL_IA_DIV(a,b) CGAL_IA_FORCE_TO_DOUBLE(CGAL_IA_STOP_CPROP(a)/CGAL_IA_STOP_CPROP(b))
+#ifdef CGAL_ALWAYS_ROUND_TO_NEAREST
+inline double CGAL_IA_UP(double d)
+{
+  // In round-to-nearest mode we find the successor instead.
+  // This preserves the interval invariants, but is more
+  // expensive and conservative.
+  return nextafter(d, std::numeric_limits<double>::infinity());
+}
+#else
+inline double CGAL_IA_UP(double d)
+{
+  // In round-upward mode we can rely on the hardware
+  // to do the job.
+  return CGAL_IA_FORCE_TO_DOUBLE(d);
+}
+#endif
+#define CGAL_IA_ADD(a,b) CGAL_IA_UP((a)+CGAL_IA_STOP_CPROP(b))
+#define CGAL_IA_SUB(a,b) CGAL_IA_UP(CGAL_IA_STOP_CPROP(a)-(b))
+#define CGAL_IA_MUL(a,b) CGAL_IA_UP(CGAL_IA_STOP_CPROP(a)*CGAL_IA_STOP_CPROP(b))
+#define CGAL_IA_DIV(a,b) CGAL_IA_UP(CGAL_IA_STOP_CPROP(a)/CGAL_IA_STOP_CPROP(b))
 inline double CGAL_IA_SQUARE(double a){
   double b = CGAL_IA_STOP_CPROP(a); // only once
-  return CGAL_IA_FORCE_TO_DOUBLE(b*b);
+  return CGAL_IA_UP(b*b);
 }
 #define CGAL_IA_SQRT(a) \
-        CGAL_IA_FORCE_TO_DOUBLE(CGAL_BUG_SQRT(CGAL_IA_STOP_CPROP(a)))
+        CGAL_IA_UP(CGAL_BUG_SQRT(CGAL_IA_STOP_CPROP(a)))
```

Comment:
It may be slightly confusing that in the new mode CGAL_IA_SQRT means up(sqrt) while in the old mode, it means sqrt with dynamic rounding, and it is actually used once with downward rounding (the other operations are only ever used with upward rounding).

Reply:
I've abstracted IA_sqrt_up and IA_sqrt_toward_zero.

It's a little awkward because of scoping.

Let me know what you think.